Add support for a new ProjectShimFactory to be consumed by CPS.

mavasani opened the pull request:
1. IProjectShim and IProjectShimFactory are the primary interfaces being introduced that CPS will use to forward project system changes to workspace hosts.

2. Refactored the UpdateOptions code in AbstractRoslynProject and sub-types so that it can be shared with the new project shim.

3. Added IVT to CPS binaries to enable consuming these APIs.

mavasani commented:
test vsi please

mavasani commented:
Tagging @davkean @srivatsn for a quick upfront review.

mavasani commented:
@dotnet-bot test vsi please

mavasani commented:
Added IVT and fixed build.

mavasani commented:
@dotnet-bot test vsi please

mavasani commented:
@dotnet-bot retest windows_vsi_p1_open_prtest please
// Retest reason: ibcmerge failed

mavasani commented:
@dotnet-bot retest windows_vsi_p1_open_prtest please
// Retest reason: https://github.com/dotnet/roslyn/issues/10788
// Previous failure: http://dotnet-ci.cloudapp.net/job/Private/job/dotnet_roslyn-internal/job/master/job/windows_vsi_p1_open_prtest/255/

mavasani commented:
@dotnet-bot retest windows_vsi_p1_open_prtest please

mavasani commented:
@dotnet-bot test vsi please

mavasani commented:
This PR is now ready for review.

Tagging @dotnet/roslyn-ide @dotnet/project-system for review.

srivatsn commented:
Tagging @mattwar as well.

mavasani commented:
@dotnet-bot test vsi please

CyrusNajmabadi commented on src/VisualStudio/CSharp/Impl/ProjectSystemShim/CSharpProjectShim.cs:
```diff
@@ -92,8 +87,12 @@ private string GetIdForErrorCode(int errorCode)
+        protected override CompilationOptions GetCompilationOptions(ParseOptions newParseOptions)
         {
+            // Get the base options from command line arguments + common workspace defaults.
+            var options = (CSharpCompilationOptions)base.GetCompilationOptions(newParseOptions);
```
You could make this type paramtereized on the type of the compilation options it has.  then you won't need these types of casts in the code.

CyrusNajmabadi commented on src/VisualStudio/CSharp/Impl/ProjectSystemShim/CSharpProjectShim.cs:
```diff
@@ -289,8 +259,10 @@ private string GetStringOption(CompilerOptions optionID, string defaultValue)
+        protected override ParseOptions GetParseOptions()
         {
+            // Get the base parse options and override the defaults with the options from state.
+            var options = (CSharpParseOptions)base.GetParseOptions();
```
Same with thsi.

CyrusNajmabadi commented on src/VisualStudio/CSharp/Test/ProjectSystemShim/CPS/CSharpReferencesTests_CPS.cs:
```diff
@@ -0,0 +1,81 @@
+    public partial class CSharpReferenceTests
+    {
+        [Fact]
+        [Trait(Traits.Feature, Traits.Features.ProjectSystemShims)]
```
all these attributes can be merged onto a single line.

mavasani replied:
I cloned these from existing ProjectShimTests, which also have the same style, I'll retain it for consistency.

CyrusNajmabadi commented on src/VisualStudio/CSharp/Test/ProjectSystemShim/CSharpHelpers.cs:
```diff
@@ -29,6 +34,62 @@ public static CSharpProjectShimWithServices CreateCSharpProject(TestEnvironment
+            string outputPath = Path.GetTempPath();
+            foreach (var arg in commandLineArguments)
+            {
+                var outPrefix = "out:";
```
const.

mattwar commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IProjectShim.cs:
```diff
@@ -0,0 +1,34 @@
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    internal interface IProjectShim : IAnalyzerHost
```
Why is this called a ProjectShim?  It seems like there should be a much better name that indicates what this interface is for.  Why does it derive from IAnalyzerHost?

davkean replied:
Agreed. We previously called this IWorkspaceContext and IWorkspaceContextFactory, but open to any name.

mavasani replied:
The type implementing this interface is a very thin layer on top of AbstractProject for forwarding project system events/properties to the workspace host. The subtypes for current C# and VB project systems are called CSharpProjectShim and VisualBasicProject, so I thought calling it a _shim_ would keep it consistent. I personally don't like WorkspaceContext as that doesn't convey that this is the _project system_ representation of the _Project_ type that exists at the Workspaces layer. I thought of "IProjectSystemProject" but wasn't sure about it. Did you have a better name suggestion?

davkean replied:
From our standpoint, *we are* the project system, Roslyn is not a project system.

mavasani replied:
> Why does it derive from IAnalyzerHost?

Basically all the APIs on IAnalyzerHost are needed to be invoked by CPS as they deal with items (ruleset, additional files, and analyzer references). Alternatively, we can avoid deriving the interface from IAnalyzerHost and have CPS explicitly MEF import IAnalyzerHost to forward analyzer host related item changes.

mattwar commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.StandardTextDocument.cs:
```diff
@@ -49,7 +49,7 @@ private class StandardTextDocument : ForegroundThreadAffinitizedObject, IVisualS
                 IVisualStudioHostProject project,
                 DocumentKey documentKey,
-                uint itemId,
+                Func<IReadOnlyList<string>> getFolderNames,
```
Why don't you just pass in the list instead of a function.

davkean commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IProjectShim.cs:
```diff
@@ -0,0 +1,34 @@
+        // Project property changes.
+        void SetProjectGuid(Guid guid);
+        void SetProjectFilePath(string projectFilePath);
+        void SetIsWebsiteProject();
```
I looked at this, these kind of behavior differences should be driven off capabilities that describe the behavior difference, not specific knowledge about specific project types. Even still, I had a look at we'll never set this, so is not needed.

mavasani replied:
Ok, I can remove this.

davkean commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/ProjectShimFactory.ProjectShim.cs:
```diff
@@ -0,0 +1,61 @@
+
+            private static string GetProjectDisplayName(string projectFilePath)
+            {
+                return PathUtilities.GetFileName(projectFilePath, includeExtension: false);
```
We need to control the display name - otherwise, you're going to end up with the same display name for all the IProjectShim instances we instantiate for one project (think cross-targeting feature that we're implementing, where we'll fire up one for every target framework)

mavasani replied:
Ok, I can avoid computing display name from project file path and make it an explicit parameter that CPS passes in.

davkean commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/ProjectShimFactory.cs:
```diff
@@ -0,0 +1,88 @@
+
+        IProjectShim IProjectShimFactory.CreateProjectShim(string languageName, string projectFilePath, Guid projectGuid, object hostObject, CommandLineArguments commandLineArguments)
+        {
+            var vsHierarchy = hostObject as IVsHierarchy;
```
Why isn't there a little more validation to help the caller here, when:

1) The language is not recognized
2) You pass something other than IVsHierarchy as the "hostObject" for now

mavasani replied:
I can add those checks and throw if not satisfied.

davkean commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs:
```diff
@@ -1322,33 +1344,53 @@ protected void SetOutputPathAndRelatedData(string objOutputPath)
+            
+            if (newPath != null && File.Exists(newPath) && _filePathOpt != newPath)
             {
                 Debug.Assert(PathUtilities.IsAbsolute(newPath));
```
Why is that the path that we pass you enforced up-front to be absolute in SetProjectFilePath so that we don't later down the track fail?

mavasani replied:
I didn't want to change the existing behavior of AbstractProject, we can probably add this validation in our sub-type of AbstractProject and handle it early.

davkean commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject_IAnalyzerHost.cs:
```diff
@@ -24,10 +25,17 @@ public void AddAnalyzerAssembly(string analyzerAssemblyFullPath)
+            if (_visualStudioWorkspaceOpt == null)
+            {
+                // This can happen only in tests.
+                var testAnalyzer = new VisualStudioAnalyzer(analyzerAssemblyFullPath, fileChangeService, this.HostDiagnosticUpdateSource, this.Id, this.Workspace, loader: null, language: this.Language);
```
If I understand this code correctly, you're solely doing this because you can't provide an IAnalyzerService? Why don't you provide a mocked one?

mavasani replied:
There are lots of places in ProjectSystemShim testing where we can add more mocking and avoid null checks. AbstractProject already has checks for null workspace at lot of places. My goal with this change was to add enough mocking so we can test the interface APIs. For example, for AddAnalyzerReference, we want to test that we can add an analyzer reference and verify it was added, without necessary attempting to verify that diagnostic analyzers within it get correctly loaded up. The later would require mocking up all the analyzer related services, but doesn't gain us any additional testing surface for our newly added interface.

davkean commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractRoslynProject.cs:
```diff
@@ -24,8 +29,13 @@ internal abstract partial class AbstractRoslynProject : AbstractProject, ICompil
-            HostDiagnosticUpdateSource hostDiagnosticUpdateSourceOpt)
-            : base(projectTracker, reportExternalErrorCreatorOpt, projectSystemName, hierarchy, language, serviceProvider, visualStudioWorkspaceOpt, hostDiagnosticUpdateSourceOpt)
+            HostDiagnosticUpdateSource hostDiagnosticUpdateSourceOpt,
+            string projectFilePath = null,
```
Looking through this code in AbstractRoslynProject, none of it seems to be hardened to be called on a background thread. Are we should reusing the existing base is a good way to approach this?

davkean replied:
Actually, I'm leaning towards that this is the wrong approach, AbstractProject is doing a bunch of work that expects its on the UI thread.

davkean commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractRoslynProject.cs:
```diff
@@ -50,33 +60,119 @@ protected CommandLineArguments GetParsedCommandLineArguments()
+            }
+
+            // Explicitly disable concurrent build.
+            options = options.WithConcurrentBuild(concurrent: false);
```
Why?

davkean commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/ProjectShimFactory.ProjectShim.cs:
```diff
@@ -0,0 +1,61 @@
+            public ProjectShim(
+                CommandLineArguments commandLineArguments,
+                VisualStudioProjectTracker projectTracker,
+                Func<ProjectId, IVsReportExternalErrors> reportExternalErrorCreatorOpt,
```
What's the hungarian notation "opt"?

davkean commented:
The more I think about this, the more I think that deriving from AbstractRoslynProject is the wrong approach for the long term:

1) It's not thread-safe - we're calling CreateProjectShim on a background thread, and this thing is calling IVsHierarchy methods, and calling services, etc on a UI thread.
2) It has behavior that doesn't make sense in CPS, if we're pushing you things such as project files, display names, guids, then why is it still listening to hierarchy events?

I'd prefer where you introduce a new version of AbstractProject that is free from that - and push behavior into it to be shared with AbstractRoslynProject/existing AbstractProject where it makes sense. This will also push us closer to having the ability to fire up a language service agnostic of VS for both VSCode and other hosts.

@jasonmalinowski @mattwar @CyrusNajmabadi Thoughts?

CyrusNajmabadi commented:
I'd like a meeting about this to discuss what's going on here.  I'm coming in 100% blind.  While i'm mostly familiar with how Roslyn does it's project systemy goo, i have no idea how CPS works.  So it's entirely unclear to me what we're trying to do here and why it's the appropriate path forward.

A meeting that explains what the goals are, what the designs of hte different systems are, and how we intend to integrate them, would go a *long* way to helping me out here.

mavasani commented:
@davkean 
I think there are couple of big disadvantages of building up a parallel AbstractProject for CPS:

1. Large pieces of code in AbstractProject are common to CPS and current project system. We'll end up either maintaining forks or putting these in shared helpers. I don't like either approach given that for Dev15 we are going to ship both the project systems.
2. This will enforce cascading changes in other components in this layer consuming AbstractProject. For example, the VisualStudioProjectTracker now needs to be changed to move off AbstractProject.

I feel a better approach is to retain _AbstractProject_ as the common base type for all project systems and abstract out the pieces where it currently violates the requirements of any of the supported project systems. We first need to jointly agree on our design goals for Dev15 versus post Dev15. I believe these to be:

1. Dev15:
  1. AbstractProject supports 2 project systems: CPS and current csproj/msvbprj based project system.
  2. Both of these are still based off IVSHierarchy for all UI centric operations, where it is fine to access the hierarchy if we are already on the UI thread.
  3. For code that can be invoked from the background thread by either project system, we need to abstract it out in AbstractProject for the sub-types to implement it. For example, fetching the properties of project/project items differs between CPS and native project system, and so we can have AbstractProject take an IPropertiesProvider to get all such properties. The CPS project system shim can implement this in non-UI fashion, while the existing project system shim can keep using the IVSHierarchy for this.
  4. For code that is specific to current project system, example listening to IVSHierarchyEvents, it should just be moved to a new sub-type of AbstractProject that current C# and VB project shims derive from.
  5. Harden AbstractProject so that sub-types can optionally enforce AssertIsForeground for Hierarchy access.

2. Post-Dev15 (when we have retired csproj/msvbproj):
  1. AbstractProject supports CPS based project systems in a VS-independent way. This can include 3rd party project system implementations, our own VS based implementation and even non-VS based implementations.
  2. Replace IVSHierarchy with an equivalent non-VS interface for all UI accesses to hierarchy for UI centric operations.
  3. Harden AbstractProject to always enforce UI thread accesses to hierarchy for all sub-types.

  I feel attempting to do any of this work in Dev15 timeframe will add to the cost for something that is not strictly required for our Dev15 scope.

mavasani commented:
@dotnet-bot retest microbuild_prtest please
// Retest reason: _error MSB3073: The command "D:\j\workspace\microbuild_pr---d013602c\src\Tools\MicroBuild\..\..\..\Binaries\Release\SignRoslyn\SignRoslyn.exe -test -binariesPath "D:\j\workspace\microbuild_pr---d013602c\src\Tools\MicroBuild\..\..\..\Binaries\Release"" exited with code 1._

davkean commented:
Why don't we use our team meeting at 1 to discuss this - and invite interested parties?

mavasani commented:
@dotnet-bot test vsi please

mavasani commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs:
```diff
@@ -159,74 +151,16 @@ internal abstract partial class AbstractProject : IVisualStudioHostProject
                 _externalErrorReporter = reportExternalErrorCreatorOpt(_id);
             }
 
-            ConnectHierarchyEvents();
```
Moved to sub-type AbstractLegacyProject

mavasani commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs:
```diff
@@ -328,7 +288,7 @@ public ProjectInfo CreateProjectInfoForCurrentState()
                 analyzerReferences: _analyzers.Values.Select(a => a.GetReference()),
                 additionalDocuments: _additionalDocuments.Values.Select(d => d.GetInitialState()));
 
-            return info.WithHasAllInformation(_intellisenseBuildSucceeded);
```
@jasonmalinowski @davkean `_intellisenseBuildSucceeded` is set by [IIntellisenseBuildTarget.SetIntellisenseBuildResult](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/AbstractProject_IIntellisenseBuildTarget.cs,25), and used to create the ProjectInfo for the project. This interface is specific to native project system, so I have moved its implementation to AbstractLegacyProject and abstracted the bool with `DesignTimeBuildStatus`. We will probably need an API on IProjectContext to set this flag - for now, its always set to true for CPS projects.

mavasani commented:
Ping, can someone please take another look? I have addressed the feedback from the meeting.

mattwar commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IProjectContext.cs:
```diff
@@ -0,0 +1,34 @@
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    internal interface IProjectContext
```
Shouldn't these API's just be members of IVisualStudioHostProject?

mavasani replied:
Possibly, but I think that interface is primarily for consumers of the project/document at VS layer, and this one is for the project system pushing down the changes. Eventually, when we make the APIs public in this layer, we should revisit the interfaces.

mattwar replied:
But this is only used by an Rolsyn specific adapter in the project system to talk to roslyn host projects, so it still works.

You only really have two choices. Either the interface is a project system interface for pushing information that the project system knows to consumers or it is a Roslyn interface for manipulating a Roslyn type.  All other choices are half measures that cause the meaning of the type to become murky.  

---
In reply to: [71422154](https://github.com/dotnet/roslyn/pull/12440#discussion_r71422154) [](ancestors = 71422154)

mavasani replied:
Our intention is the first approach "project system interface for pushing information that the project system knows to consumers"

mattwar replied:
Then you probably don't want the ProjectId property and other elements that tie it to Roslyn.

---
In reply to: [71427489](https://github.com/dotnet/roslyn/pull/12440#discussion_r71427489) [](ancestors = 71427489)

mavasani replied:
Makes sense. I can probably change the Add/RemoveProjectReference APIs below to just take the IProjectContext.

mavasani replied:
@davkean Do we want the `SetCommandLineArguments` to also pass in a string instead of parsed command line arguments?

jasonmalinowski replied:
I believe yes: otherwise who is doing the parsing? We probably want the ability for the language service + compilers to be updated for a new switch, and not having to require the project system to rev too, at least for day-to-day development.

mavasani replied:
That's a good point, I'll change this over to take a string, unless @davkean has a strong reason not to.

jasonmalinowski replied:
Actually I take that back: as long as you're the ones calling parse in proc, then our binding redirects means you're picking up our compiler. So it probably doesn't matter. I defer to your preference.

jasonmalinowski commented on src/VisualStudio/CSharp/Test/ProjectSystemShim/CPS/AnalyzersTests_CPS.cs:
```diff
@@ -0,0 +1,80 @@
+
+namespace Roslyn.VisualStudio.CSharp.UnitTests.ProjectSystemShim
+{
+    public partial class AnalyzersTests
```
Please keep file names matching class names. Just using namespaces to distinguish seems sufficient to me.

mavasani replied:
Sure, will do.

jasonmalinowski commented on src/VisualStudio/CSharp/Test/ProjectSystemShim/CPS/CSharpReferencesTests_CPS.cs:
```diff
@@ -0,0 +1,81 @@
+                var project1 = (AbstractProject)CreateCSharpCPSProject(environment, "project1", commandLineArguments: @"/out:c:\project1.dll");
+                var project2 = (AbstractProject)CreateCSharpCPSProject(environment, "project2", commandLineArguments: @"/out:c:\project2.dll");
+                var project3 = (AbstractProject)CreateCSharpCPSProject(environment, "project3", commandLineArguments: @"/out:c:\project3.dll");
+                var project3Shim = (IProjectContext)project3;
```
Why all the casting here?

mavasani replied:
I'll clean this up.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IProjectContextFactory.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    internal interface IProjectContextFactory
+    {
+        IProjectContext CreateProjectContext(string languageName, string projectDisplayName, string projectFilePath, Guid projectGuid, string projectTypeGuid, object hostObject, CommandLineArguments commandLineArguments);
```
I thought we were getting rid of GUIDs? Or is CPS still making GUIDs, just on the fly each time the project is opened? Does it ensure any stability of those GUIDs?

jasonmalinowski replied:
Why is projectTypeGuid a string?

jasonmalinowski replied:
In general, full doc comments on this method might be important, given this is core to the whole system.

srivatsn replied:
The solution creates the project guids and calls the project system to set it. We will just hold onto it in memory instead of setting it. The guid will still be persisted in the sln file.

mavasani replied:
We are getting rid of guids from the project file, they'll still be associated with projects and stay in the solution file. I have added the GUID parameter to this method as AbstractProject sets it in the ctor and I didn't want to change those semantics. We can get retain get rid of GUID/projectType guid parameters to this method if we feel CPS won't set them here. I already added methods on IProjectContext to set these explicitly at a later time.

mavasani replied:
> Why is projectTypeGuid a string?

It was already a string field in AbstractProject, I have retained it as is.

jasonmalinowski replied:
Ah, it looks like the type guid is being used for telemetry reasons only; you aren't having an equivalent, right? Then move that back to the legacy layer.

mavasani replied:
We are surely not going to have project type guids in project file, but I am not sure if they are not there even in memory/sln file - @davkean can probably answer that.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/ProjectContextFactory.cs:
```diff
@@ -0,0 +1,88 @@
+
+        IProjectContext IProjectContextFactory.CreateProjectContext(string languageName, string projectDisplayName, string projectFilePath, Guid projectGuid, string projectTypeGuid, object hostObject, CommandLineArguments commandLineArguments)
+        {
+            var vsHierarchy = hostObject as IVsHierarchy;
```
I don't get this bit. If we are demanding a hierarchy, then why isn't it strongly typed as such?

mavasani replied:
Dave wanted this to be an object so that in future when we support a host object that is not necessarily a IVSHierarchy, we don't need to change the primary factory method. I am fine with your view here, we can probably just strongly type this for clarity now and deal with API change when we do add such support (probably can just add a new overload with new type of host object instead of "object" type).

jasonmalinowski replied:
Indeed, I'd rather we declare our warts than try to hide them. Using 'object' might mean we can avoid a API break in the signature, but since the code on both sides is still strongly coupled on both sides. Also, as long as the RDT still talks in hierarchies as it's currency, I'm not sure we'll be able to remove that. I can think of some other API shapes which might make this more strongly typed, but I still forsee the parameter here being the least of our worries when such a refactoring would happen.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/LegacyProjectSystem/AbstractLegacyProject.cs:
```diff
@@ -0,0 +1,354 @@
+        /// <summary>
+        /// Returns a display name for the given project.
+        /// </summary>
+        private static bool TryGetProjectDisplayName(IVsHierarchy hierarchy, out string name)
```
This method is doing nothing other than forward to .TryGetName...?

mavasani replied:
Will clean it up, don't need 2 methods here.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/Venus/ContainedDocument.cs:
```diff
@@ -117,7 +117,7 @@ internal sealed class ContainedDocument : ForegroundThreadAffinitizedObject, IVi
             this.Key = new DocumentKey(Project, filePath);
             this.Id = DocumentId.CreateNewId(Project.Id, filePath);
-            this.Folders = containedLanguage.Project.GetFolderNames(itemId);
+            this.Folders = (containedLanguage.Project as AbstractLegacyProject)?.GetFolderNames(itemId);
```
Are you trying to support ContainedDocuments in non-legacy scenarios? Otherwise I think I'd prefer a hard cast; this might just flow strange data around and it'll be hard to debug.

mavasani replied:
I believe Venus is not in plan for Dev15 for CPS, though @davkean @srivatsn can confirm. I'll change it to a hard cast.

jasonmalinowski commented on src/VisualStudio/CSharp/Impl/ProjectSystemShim/CSharpProjectShim.cs:
```diff
@@ -302,12 +274,10 @@ protected CSharpParseOptions CreateParseOptions()
-                preprocessorSymbols: symbols.AsImmutable(),
-                documentationMode: documentationMode)
-                .WithFeatures(GetParsedCommandLineArguments().ParseOptions.Features);
+            return options.WithKind(SourceCodeKind.Regular)
```
We generally avoided doing all the .With calls to avoid extra garbage creation, if I recall.

mavasani replied:
I think that holds if we are creating options from scratch. As the default workspace options match for current AbstractLegacyProject and CPS based projects, I have moved that code down to the AbstractProject.GetCompilationOptions(), which necessitates the sub-types to modify the base options with .With methods.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs:
```diff
@@ -1007,6 +970,8 @@ public virtual void Disconnect()
 
                 _pushingChangesToWorkspaceHosts = false;
+
+                this.EditAndContinueImplOpt = null;
```
Does this need to be closed/disposed/something else?

mavasani replied:
I have retained the [current references/assignments](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/AbstractRoslynProject.cs,1b75062494c3b88d,references) to it, just moved it out from AbstractRoslynProject down to AbstractProject.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IProjectContext.cs:
```diff
@@ -0,0 +1,34 @@
+        // Project property changes.
+        void SetProjectGuid(Guid guid);
+        void SetProjectTypeGuid(string projectTypeGuid);
+        void SetProjectDisplayName(string projectDisplayName);
```
Why not just make the properties read/write?

mavasani replied:
That's another option, not sure if CPS will care about reading them though. Dave explicitly requested DisplayName and ProjectFilePath to be readable for CPS state tracking.

jasonmalinowski replied:
I guess I have no problem with them being read and write, my problem is why there's readonly properties with setter methods. :smile:

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/ProjectContextFactory.ProjectContext_IProjectContext.cs:
```diff
@@ -0,0 +1,95 @@
+            {
+                var referencedProject = this.ProjectTracker.GetProject(projectId);
+                var binPathOpt = referencedProject.TryGetBinOutputPath();
+                if (!string.IsNullOrEmpty(binPathOpt))
```
Why if there's not an output path? Please comment why.

mavasani replied:
AbstractProject and ProjectTracker is tracking project references through bin paths. I'll add a comment.

jasonmalinowski replied:
Did the legacy shims do that?

mavasani replied:
LegacyShims always added project references by filePath, see [here](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices.CSharp/ProjectSystemShim/CSharpProjectShim.ICSharpProjectSite.cs,106) and [here](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/AbstractProject.cs,478)

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/ProjectContextFactory.ProjectContext_IProjectContext.cs:
```diff
@@ -0,0 +1,95 @@
+                }
+            }
+
+            public void RemoveProjectReference(ProjectId projectId)
```
We decided that the scenario of project A referencing project B with more than one reference (but different properties) wasn't supported? I suspect it wasn't, just wanted to make sure.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs:
```diff
@@ -31,7 +30,7 @@ internal abstract partial class AbstractProject : IVisualStudioHostProject
         private readonly ProjectId _id;
         private readonly string _language;
-        private readonly IVsHierarchy _hierarchy;
+        private readonly IVsHierarchy _hierarchyDoNotAccessDirectly;
```
Why not, since it's readonly and therefore not mutating or anything? Should this just become an autoprop like you did for some others?

mavasani replied:
Yes, should be an autoprop.

jasonmalinowski replied:
While you're here, just to do the others too.

jasonmalinowski commented on src/VisualStudio/CSharp/Test/ProjectSystemShim/AnalyzersTests.cs:
```diff
@@ -12,7 +12,7 @@
 namespace Roslyn.VisualStudio.CSharp.UnitTests.ProjectSystemShim
 {
-    public class AnalyzersTests
+    public partial class AnalyzersTests
```
Eww, partial across folders. :-(

mavasani replied:
Yes, need to change it to a separate class with a different namespace.

jasonmalinowski commented on src/VisualStudio/CSharp/Test/ProjectSystemShim/CSharpHelpers.cs:
```diff
@@ -29,6 +33,64 @@ public static CSharpProjectShimWithServices CreateCSharpProject(TestEnvironment
+                parsedArguments);
+        }
+
+        public static void SetCommandLineArguments(IProjectContext cpsProject, params string[] commandLineArguments)
```
Parameters could use some updated names.

mavasani replied:
Yes sure.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/LegacyProjectSystem/AbstractLegacyProject.cs:
```diff
@@ -0,0 +1,354 @@
+using Roslyn.Utilities;
+using VSLangProj;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
```
Please update namespaces to match the folder names.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/ProjectContextFactory.ProjectContext.cs:
```diff
@@ -0,0 +1,66 @@
+                projectTracker.AddProject(this);
+            }
+
+            protected override CommandLineArguments ParseCommandLineArguments(IEnumerable<string> arguments)
```
This method seems fishy: this is used for ICompilerOptionsHostObject implementation, which is only for the legacy system. I suspect you can move the code down to AbstractLegacyProject.

mavasani replied:
Yes, in its current shape, unless we decide to change the CPS SetCommandLineArguments API to take a string.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/ProjectContextFactory.ProjectContext.cs:
```diff
@@ -0,0 +1,66 @@
+            {
+                base.PostSetOptions();
+
+                // If outputPath has changed, then invoke SetOutputPathAndRelatedData to update the project tracker bin path for this project.
```
Why not just do this in the base type? I think that eliminates the whole PostSetOptions thing?

mavasani replied:
LegacyShims already know when this method needs to be explicitly invoked, see [here](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices.CSharp/ProjectSystemShim/CSharpProjectShim.ICSInputSet.cs,48). CPS doesn't understand this so we need to explicitly check if the output path was changed with the new command line arguments, and invoke it here.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject_Options.cs:
```diff
@@ -0,0 +1,134 @@
+
+        protected void SetCommandLineArguments(CommandLineArguments commandLineArguments)
+        {
+            _lastParsedCommandLineArguments = commandLineArguments;
```
This doesn't seem right; this variable is supposed to track the last parsed thing of _lastParsedCompilerOptions, or at least that's what it used to do.

mavasani replied:
I'll clean this up once @davkean finalizes if we need to pass in command line as a string (_lastParsedCompilerOptions would stay in AbstractProject) or as CommandLineArguments (_lastParsedCompilerOptions and ParseCommandLineArguments will move into AbstractLegacyProject)

jasonmalinowski commented:
So I really don't get how CompilationOptions/ParseOptions work any more. The AbstractProject implementation seems to mix CPS demands and legacy demands in a really unclear way. This wasn't a clean area of the codebase before, but I _really_ don't understand it now.

I would have expected the contract between AbstractProject and the derived types was still two methods of setting the ParseOptions and CompilationOptions types directly, abstracting the underlying source of them. AbstractLegacyProject would have all the guts that already existed. The CPS system then would do the much simpler code with parsing, ultimately arriving at those two types and then just passing those down that way.

jasonmalinowski commented:
Also, I would request XML doc comments on the newly introduced types, that should mean that I could hand this code to somebody who wasn't in our team room conversation and they could understand what the newly introduced types mean and what their relationship in the type hierarchy is.

mavasani commented:
> So I really don't get how CompilationOptions/ParseOptions work any more

Basically, previous code in LegacyShims to compute options was fetching parts of options from different sources:

1. Use default parsed CompilationOptions from the CommandLineParser (values were same for VB and C#, but duplicated in both layers): e.g. [here](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices.CSharp/ProjectSystemShim/CSharpProjectShim.cs,224)
2. Use workspace defaults for some options (values were same for VB and C#, but duplicated in both layers): e.g. [here](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices.CSharp/ProjectSystemShim/CSharpProjectShim.cs,219)
3. Use options set explicitly via LegacyShim interfaces: e.g. [here](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices.CSharp/ProjectSystemShim/CSharpProjectShim.cs,205)

New code moves 1. and 2. down to [AbstractProject.GetCompilationOptions](https://github.com/dotnet/roslyn/pull/12440/files/98518abcb8c55a5306a7ccaf646139ae6e19f267#diff-46bfd78350e474a83268ba2e2c978f5aR81) - these options are same for VB and C# across CPS and LegacyShims. Sub-types override these to fetch these default options and then set options specific to the underlying project systems (basically 3. above).

crummel commented:
Build triggered. sha1 is merged.

mavasani commented:
@jasonmalinowski I have addressed the review feedback.

I have still retained the old signature for SetCommandLineArguments, taking parsed CommandLineArguments. If required, we can change it to a string in future.

mavasani commented:
@dotnet-bot test vsi please

mavasani commented:
@dotnet-bot retest windows_vsi_p0_open_prtest please
// Retest reason: https://github.com/dotnet/roslyn/issues/11429

mavasani commented:
@dotnet-bot retest windows_vsi_p2_open_prtest please
// Previous failure: http://dotnet-ci.cloudapp.net/job/Private/job/dotnet_roslyn-internal/job/master/job/windows_vsi_p2_open_prtest/312/testReport/
// Retest reason: Timing issue (Apex was not ready after 300 seconds.)

mavasani commented:
@dotnet-bot retest windows_vsi_p2_open_prtest please
// Retest reason: Another timing issue (http://dotnet-ci.cloudapp.net/job/Private/job/dotnet_roslyn-internal/job/master/job/windows_vsi_p2_open_prtest/313/testReport/(root)/CSharpFixAllOccurrences/FixAllOccurrences_in_Solution___Remove_this_qualification/) - I'll open a bug if it fails again.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs:
```diff
@@ -954,11 +893,8 @@ private void RemoveGeneratedDocument(IVisualStudioHostDocument document)
         public virtual void Disconnect()
         {
-            using (_visualStudioWorkspaceOpt?.Services.GetService<IGlobalOperationNotificationService>()?.Start("Disconnect Project"))
+            using (Workspace?.Services.GetService<IGlobalOperationNotificationService>()?.Start("Disconnect Project"))
```
This rename lost the opt-ness of it.

jasonmalinowski commented:
I'm still not getting the options thing. Perhaps my concern is just because there's so much marked protected I can't figure out how the data is *supposed* to flow. It seems if you're deriving from AbstractProject, you can impact the options via one of the following ways:

1. You can call AbstractProject.SetCommandLineArguments
2. You can directly set AbstractProject.ParsedCommandLineArguments.
3. You could override GetCompilationOptions.
4. You can directly call SetOptions.

Each of these have different behaviors, and I can't tell for the life of me the intent. Also, what's the relationship between ParsedCommandLineArguments and the CurrentCompilationOptions and CurrentParseOptions? Which one should somebody consume?

Previously, there was a (very opaque) data flow: the only meaningful flow into the base layer was through SetOptions. The parsing logic was shared, but that didn't actually push anything through; the core code just told the drived ones to pick up the values, and then call SetOptions.

I suspect this might get a lot cleaner if you don't try to unify the command line arguments, since they're being used for two completely different reasons. The CPS one is the source of truth; the legacy one is only a small set of options that are explicitly not the truth and you shouldn't accidentally use those values.  There was a reason we just picked up the handful of values explicitly instead of trying to reuse everything.

jasonmalinowski commented:
I should say though everything else looks really good. I just really don't get this one part!

mavasani commented:
The expected flow is: you set ParsedCommandLineArguments, then you invoke UpdateOptions, which in turn computes the options to set with 3 steps:

1. Uses the Compilation and Parse options from ParsedCommandLineArguments
2. Applies workspace specific default options and
3. Optionally allow sub-type to append its own options from some other source, by overriding AbstractProject.GetCompilationOptions/GetParseOptions.

I think there are 2 sources of confusion, which I hope below can address:

1. Difference between CurrentCompilationOptions/CurrentParseOptions versus GetCompilationOptions/GetParseOptions. The former represent the last set values for each of these respectively, while the later actually creates new options based on current state (ParsedCommandLineArguments + project specific state) - I think I should rename the later methods to CreateCompilationOptions/CreateParseOptions to make it more explicit.

2. Another confusion is coming from the fact that there are 2 ways to set ParsedCommandLineArguments, directly via the setter or via SetCommandLineArguments, which then also invokes UpdateOptions. I can clean this up to just retain the setter, and anyone who also wants to UpdateOptions can explicitly invoke it.

The expectation then is: You need to ensure that ParsedCommandLineArguments property is set to be non-null prior to invoking UpdateOptions. Sub-type is free to change its state by updating ParsedCommandLineArguments prior to invoking UpdateOptions (CPS) or update its own storage state (Legacy shims) or both. Once it has done so, invoking UpdateOptions will always create new options and apply them to the workspace.

mavasani commented:
@jasonmalinowski I have cleaned up the options infrastructure further and it is now very similar to the old flow/APIs.

1. _lastParsedCompilerOptions - belongs to AbstractLegacyProject for setting options from command line string for LegacyShim interfaces.
2. _lastParsedCommandLineArguments - belongs to AbstractProject for setting options from parsed CommandLineArguments.

Simplified UpdateOptions and CreateCompilationOptions/CreateParseOptions APIs to optionally set the above fields, create new options from the latest state, and set this options:

1. AbstractProject.UpdateOptions() - Creates options using _lastParsedCommandLineArguments and state from sub-type and sets these options.
2. AbstractProject.SetArgumentsAndUpdateOptions(CommandLineArguments) - Sets _lastParsedCommandLineArguments  and then invokes (1).
3. AbstractLegacyProject.UpdateOptions(string) - If given command line string is different from _lastParsedCompilerOptions, then parses it to get new CommandLineArguments and invokes (2). Otherwise, invokes (1) to update options with _lastParsedCommandLineArguments.

Hope this is more cleaner now...

davkean commented:
Add notes as I run into this:

1) I now feel that the command-line should be a string, otherwise, I now need to know which language to use to parse the command-line - I've already provided language name, you should be able to do it.

2) Why is projectType a string? Do you need this up-front? What exactly are you doing with it? It's feels like something that language should not know about.

3) How do I answer "isInCurrentContext"?

mavasani commented:
> 1) I now feel that the command-line should be a string or rather an array of arguments, otherwise, I now need to know which language to use to parse the command-line - I've already provided language name, you should be able to do it.

Yes, we can do it on Roslyn side, but note that is some additional work. Currently CPSProject and CPSProjectFactory implementation in Roslyn side is language-independent and this would require language specific components, MEF import, etc. stuff. I'll try to address that today, if time permits, otherwise next week. My only concern is this PR will drag on further.

> 2) Why is projectType a string? Do you need this up-front?
This is how legacy shim communicates it to Roslyn and how it is stored in AbstractProejct in the constructor. We can avoid the field in CreateProjectShim method if you don't intend to pass it upfront.

> 3) How do I answer "isInCurrentContext"?
This is the shared project thing we discussed, basically getting [this](http://source.roslyn.io/Microsoft.VisualStudio.LanguageServices/R/a3d3388af3d5074f.html) without accessing hierarchy.

davkean commented:
More notes, how do I disconnect and shut it down?

jasonmalinowski commented:
The project type thing looks like it's only being used for telemetry purposes; since you don't quite have the same context now, we should just make that legacy-only and come up with a better plan for that.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject_Options.cs:
```diff
@@ -0,0 +1,113 @@
+            UpdateOptions();            
+        }
+
+        protected void SetOptions(CompilationOptions newCompilationOptions, ParseOptions newParseOptions)
```
Looks like this is only used from UpdateOptions now above. Should this be either private or just inlined so that way nobody can call this and bypass the other code?

mavasani replied:
This is still used by VisualBasicProject (legacy VB shim) [here](https://github.com/dotnet/roslyn/pull/12440/files/7f3e7c8d7f79d6e14d6d9c649d2eef9db130495a#diff-dd53e8d47425333c43f1961e0a9df52bR493) - Same as before this change [here](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices.VisualBasic/ProjectSystemShim/VisualBasicProject.vb,484)

jasonmalinowski replied:
Oh cute, there's a bug in CodeLens!

![image](https://cloud.githubusercontent.com/assets/201340/17069059/5b8345d6-5009-11e6-97ce-a11e0da86a68.png)

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/Legacy/AbstractLegacyProject_ICompilerOptionsHostObject.cs:
```diff
@@ -0,0 +1,46 @@
+        /// Parses the given command line (if different from the last command line), and updates options using the latest state.
+        /// </summary>
+        /// <param name="commandLine">Optional command line to parse options. If null, default CommandLineArguments will be used.</param>
+        protected void UpdateOptions(string commandLine = null)
```
This is shadowing the base method, which isn't what you want. This actually means there's a (possible) bug, as a call to `UpdateOptions` will toss the command line with the string.Empty. Just move the part here dealing with _lastParsedCompilerOptions in SetCompilerOptions above, and then this method goes away.

mavasani replied:
The legacy shim sub-types still need to be able to invoke this method such that for the first time empty command line arguments are used and from then on last parsed arguments are used.

jasonmalinowski replied:
It's not though. CSharpProejctShim.ICSInputSet calls it on any incremental option change.

jasonmalinowski replied:
I'm also not sure why the default initialize case matters either. Just set the _lastParsedCompilerOptions to empty and call ParseCommandLineArguments to initialize it in the constructor, and then the problem goes away.

jasonmalinowski replied:
The bugs aside, the implicit shadowing is really terrifying. :-(

mavasani replied:
Ah got you, I can then get rid of this method..

jasonmalinowski commented:
So options cleanup is good, other than the one bug that's now there. :-)

mavasani commented:
Thanks @jasonmalinowski I'll remove the AbstractLegacyProject.UpdateOptions(string) method and that should address it..

mavasani commented:
Done - removed AbstractLegacyProject.UpdateOptions(string)

jasonmalinowski commented:
Looks great. :shipit:

tvsonar commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/CPSProjectFactory.cs:
```diff
@@ -0,0 +1,93 @@
+                    if (!_projectLangaugeToErrorCodePrefixMap.TryGetValue(languageName, out errorCodePrefix))
+                    {
+                        Debug.Fail($"Unknown language '{languageName}'");
+                        _projectLangaugeToErrorCodePrefixMap.Add(languageName, languageName);
```
@mavasani This line seems to be wrong. `_projectLangaugeToErrorCodePrefixMap` is immutable, so this line has no effect.
